[Serialization] Drop inherited conformances on classes

jrose-apple opened the pull request:
These can be recreated if needed in a client library.

Note that this doesn't drop inherited conformances from the entire serialized interface, just from the list that a class explicitly declares.

rdar://problem/50541451

jrose-apple commented:
@swift-ci Please test source compatibility

slavapestov commented on lib/Serialization/Serialization.cpp:
```diff
@@ -3142,6 +3142,9 @@ void Serializer::writeDecl(const Decl *D) {
 
     auto conformances = theClass->getLocalConformances(
                           ConformanceLookupKind::All, nullptr);
+    llvm::erase_if(conformances, [](const ProtocolConformance *conf) {
```
Should this be a new ConformanceLookupKind?

jrose-apple replied:
Maybe. This is a quick-and-dirty test to see if it breaks things, without really thinking through the implications.

jrose-apple commented:
More real version, but still needs tests.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple committed:
[Serialization] Drop inherited conformances on classes

These can be recreated if needed in a client library. To do this, I've
added a new ConformanceLookupKind::NonInherited, which can also be
used elsewhere in the project where we're already filtering out
inherited conformances some other way.

Note that this doesn't drop inherited conformances from the entire
serialized interface, just from the list that a class explicitly
declares. They still get referenced sometimes.

rdar://problem/50541451 and possibly others

jrose-apple commented:
Passed debug source compat suite, [release still going here](https://ci.swift.org/job/swift-PR-source-compat-suite/2978/)

@swift-ci Please test

jrose-apple commented:
Hm, that same thing failed on Linux twice. Maybe it's not a fluke after all.

jrose-apple commented:
Did not fail on my own Linux build today (admittedly with a simpler invocation). Let's try one more time.

@swift-ci Please test Linux
